NCM Stress Test Report

zzxgzgz opened the pull request:
This PR includes the report of NCM Stress Test.

lfu-ps reviewed (changes requested):
1. 3rd file: where is the gray area? Is it 100% so covered by blue or red ?
2. 4th file: in excel, there's a way to show numbers of the height (Y axis), it will add some level of clarity to the slope chart

zzxgzgz commented:
> 1. 3rd file: where is the gray area? Is it 100% so covered by blue or red ?
> 2. 4th file: in excel, there's a way to show numbers of the height (Y axis), it will add some level of clarity to the slope chart

1. The gray area is so small that it is not noticable in either of the graphs.
2. I added the values for the Y axis, and changed the color of the 1 VPC testcases to green.

lfu-ps reviewed (approved):
LGTM

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+We tested NCM with different numbers of VPCs, and different number of ports inside each VPC.
+
+image::ncm_stress_test_setup.png[test_setup, 800]
```
Recommended document changes:
(1) Make "Test Setup and Detailed Workflow" as a subsection of "Overview", and move Line 14 and 18 to that new subsecion.
(2) Create a new subsection called "Test Scenarios", and move Line #16 to that new subsection. Also, mention the limitation of the tests by mentioning that NCM is deployed in a single instance, and only two on-demand clients. 
(3) At the end, we could create a new section called "Future Improvement" or "Next Step", mentioning that we will test large-scale scenarios, for example, 10K clients and scaled-out NCM etc.

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+At the beginning of the test, the Test Controller first generates the GoalStateV2 messages with the number of VPCs, and number of ports inside each VPC, and it sends those GoalStateV2s to NCM.
+
+NCM, after receiving and storing these GoalStateV2s into its cache, sends down the GoalStates to the host agents, which are the Dubhe agent on Node 1 and ACA on Node 2. 
```
A natural question people may raise: why do we install two different agents, i.e., ACA and Dubhe agent, at the same time? Do we just to test functionality, or expect some performance difference from the two?

zzxgzgz replied:
That's a good question. Whatever agent on the Node 2 doesn't really matter in these tests, it could be the Dubhe Agent or ACA. That agent's role is only to be there and receive the GoalStateV2 that was created by the TC. But in the future we can definitely replace this ACA with the Dubhe agent and then find out a way to let these agents to send the on-demand requests at the same time. I will add this to the `Future Work` section.

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+== Test Setup and Detailed Workflow
+
+The test setup is very similar with our previous tests on the on-demand workflow. Except that we replced one of the Alcor-Control-Agent(ACA) on one of the nodes with the newly written Dubhe Agent, a prototype written in Golang that mimics the gRPC server/client ability of ACA. Of course, for the Test Controller, changes were made in order to support the multi-vpc test scenarios.
```
So we would expect a TC PR as well :-)?

zzxgzgz replied:
Yes. The TC PR is bigger and I will take my time to make it look nicer.

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+With the GoalStateV2 received, NCM now understands what VPCs and ports it has and what neighbors those ports have, and it is ready for begin the stress test. The user sets the duration of the test(in seconds) and the rate of sending on-demand requests(requests/second, we can call this number X) when the Dubhe agent starts, and test will be done according to these parameters.
+
+After the test, the Dubhe agent outputs the number of requests sent, the number of GoalStateV2 it received, and the number of on-demand replies it received. If these three numbers are at the same level, we would say that NCM is able to handle X number of on-demand requests per second, with this number of VPCs and ports in its cache.
```
Does "at the same level" mean the exact number of messages?

zzxgzgz replied:
Not the exact number, because when the time is up, there are usually some replies that didn't come back on time and they are missed.

I'd say that if the difference is within one or two hundreds, it is still the same level. Because, even if the on-demand rate is slightly bigger than what NCM can handle, say we set the rate to 750, but atually the NCM max QPS is 700, we will see a significant(I'll say at least 30%) drop of GoalStateV2s received and the on-demand replies received.

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+After the test, the Dubhe agent outputs the number of requests sent, the number of GoalStateV2 it received, and the number of on-demand replies it received. If these three numbers are at the same level, we would say that NCM is able to handle X number of on-demand requests per second, with this number of VPCs and ports in its cache.
+
+With the same number of VPCs and ports, we repeat the test with a increasing X, and we stop when NCM is not able to handle the requests, and we call that NCM's max QPS is X with this amount of VPCs and ports in each VPC.
```
Describe the symptoms observed in NCM when "it is not able to handle the requests",

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+
+Note: There's no test performed for total number of ports less than 1000 when there are multiple VPCs, that's why the corresponding area of the above graph is empty.
+
+image::ncm_stress_test_latencies_ms.png[latencies_ms, 800]
```
In the diagram, we didn't find the unit of Y axis. Please add it.

xieus commented on benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc:
```diff
@@ -0,0 +1,97 @@
+. The gRPC client that sends the on-demand GoalStateV2 to the host also contributes alot to the overall latency. We are able to find out that, the `onNext` function of the Java async gRPC library, which sends the message, can take a lot of time. To our understanding, the `onNext` not take that much time as it is a asynchronous and its main purpose is to put the message into a queue(https://github ...
+. On the other hand, the gRPC server that replies the on-demand requests are performing well. From the latencies charts, we can barely see it as its latency is so little.
+
+== Running the test
+
+If you wish to run the test yourself, you shall utilize the Test Controller(https://github.com/futurewei-cloud/alcor/blob/master/services/pseudo_controller/src/main/java/com/futurewei/alcor/pseudo_controller/pseudo_controller.java), the Dhube Agent(https://github.com/futurewei-cloud/dubhe_agent) and the ACA(https://github.com/futurewei-cloud/alcor-control-agent/). If you would like to produce a t ...
+
```
Like this section. It is super 👍

zzxgzgz replied:
Thank you.

xieus reviewed (commented):
@zzxgzgz @lfu-ps This is a great report overall. Left a few minor comments.

zzxgzgz reviewed (commented):
Addressed the issues brought up by reviewers.

xieus reviewed (approved):
Good first step to get a more high-performance NCM.
